File: backend/models/account_archive.py
"""Strict response contracts for the local-first music archive."""

from __future__ import annotations

from typing import Literal, Union

from pydantic import BaseModel, ConfigDict, Field

ArchiveDataStatus = Literal["ready", "partial", "empty"]
ArchiveCapabilityStatus = Literal["available", "partial", "unavailable"]
ArchiveFeatureRole = Literal[
    "first_saved",
    "latest_saved",
    "oldest_release",
    "newest_release",
]
ArchiveChapterStatus = Literal["available", "partial", "unavailable"]
ArchiveMetricDisplayStatus = Literal["stable_rate", "count_only", "unavailable"]


class StrictArchiveModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class ArchiveCounts(StrictArchiveModel):
    saved_tracks: int = Field(ge=0)
    saved_albums: int = Field(ge=0)
    saved_artists: int = Field(ge=0)
    saved_shows: int = Field(ge=0)
    playlists: int = Field(ge=0)
    playlist_items: int = Field(ge=0)


class ArchiveCoverage(StrictArchiveModel):
    saved_tracks_with_date: int = Field(ge=0)
    saved_tracks_with_date_pct: float = Field(ge=0, le=100)
    saved_tracks_linked_to_history: int = Field(ge=0)
    saved_tracks_linked_to_history_pct: float = Field(ge=0, le=100)
    saved_tracks_with_known_duration: int = Field(ge=0)
    saved_tracks_with_known_duration_pct: float = Field(ge=0, le=100)
    known_duration_ms: int = Field(ge=0)


class ArchivePeriod(StrictArchiveModel):
    first_saved_at: str | None = None
    latest_saved_at: str | None = None
    first_play_date: str | None = None
    latest_play_date: str | None = None


class ArchiveDateProvenance(StrictArchiveModel):
    oauth: int = Field(ge=0)
    manual: int = Field(ge=0)
    legacy: int = Field(ge=0)
    missing: int = Field(ge=0)


class ArchiveCapabilities(StrictArchiveModel):
    collection_browse: ArchiveCapabilityStatus
    collection_timeline: ArchiveCapabilityStatus
    playback_cross_analysis: ArchiveCapabilityStatus


class ArchiveFeaturedItem(StrictArchiveModel):
    role: ArchiveFeatureRole
    track_name: str
    artist_name: str
    album_name: str | None = None
    added_date: str | None = None
    release_date: str | None = None
    cover_url: str | None = None
    deep_link: str | None = None


class ArchiveOverviewResponse(StrictArchiveModel):
    schema_version: Literal["account_archive_v1"] = "account_archive_v1"
    content_version: Literal["account_archive_v1_0"] = "account_archive_v1_0"
    data_revision: str
    status: ArchiveDataStatus
    counts: ArchiveCounts
    coverage: ArchiveCoverage
    period: ArchivePeriod
    date_provenance: ArchiveDateProvenance
    capabilities: ArchiveCapabilities
    featured_items: list[ArchiveFeaturedItem] = Field(default_factory=list, max_length=4)


class ArchiveFilterContext(StrictArchiveModel):
    context_version: Literal["account_archive_filter_v1"] = "account_archive_filter_v1"
    min_ms: int = Field(ge=0)
    music_only: Literal[True] = True
    merge_enabled: bool
    dynamic_threshold: bool
    max_merge_gap_minutes: int | None = Field(default=None, ge=1, le=240)
    merge_level: int = Field(ge=1, le=3)
    timezone: Literal["Asia/Shanghai"] = "Asia/Shanghai"
    first_play_at: str | None = None
    latest_play_at: str | None = None
    latest_play_date: str | None = None
    source_revision: str
    track_group_revision: str
    filter_fingerprint: str


class ArchiveGrowthPoint(StrictArchiveModel):
    period: str
    year: int = Field(ge=1900)
    quarter: int | None = Field(default=None, ge=1, le=4)
    saved_tracks: int = Field(ge=0)
    cumulative_saved_tracks: int = Field(ge=0)


class ArchiveJourneyCoverage(StrictArchiveModel):
    saved_tracks: int = Field(ge=0)
    saved_tracks_with_date: int = Field(ge=0)
    invalid_added_dates: int = Field(ge=0)
    saved_tracks_with_known_duration: int = Field(ge=0)
    duration_coverage_pct: float = Field(ge=0, le=100)


class ArchiveDurationFacts(StrictArchiveModel):
    known_duration_ms: int = Field(ge=0)
    release_year_start: int | None = None
    release_year_end: int | None = None


class ArchiveJourneyResponse(StrictArchiveModel):
    schema_version: Literal["account_archive_journey_v1"] = "account_archive_journey_v1"
    content_version: Literal["account_archive_journey_v1_0"] = "account_archive_journey_v1_0"
    data_revision: str
    status: ArchiveChapterStatus
    filter_context: ArchiveFilterContext
    coverage: ArchiveJourneyCoverage
    duration: ArchiveDurationFacts
    annual_growth: list[ArchiveGrowthPoint] = Field(default_factory=list)
    quarterly_growth: list[ArchiveGrowthPoint] = Field(default_factory=list)
    milestones: list[ArchiveFeaturedItem] = Field(default_factory=list, max_length=2)


class ArchiveEntityPreview(StrictArchiveModel):
    track_name: str
    artist_name: str
    album_name: str | None = None
    cover_url: str | None = None
    deep_link: str | None = None
    added_date: str | None = None
    first_play_at: str | None = None
    last_play_at: str | None = None
    effective_plays: int = Field(default=0, ge=0)


class ArchiveCohortCoverage(StrictArchiveModel):
    saved_tracks: int = Field(ge=0)
    matched_saved_tracks: int = Field(ge=0)
    unmatched_saved_tracks: int = Field(ge=0)
    canonical_saved_entities: int = Field(ge=0)
    dated_canonical_entities: int = Field(ge=0)
    invalid_added_dates: int = Field(ge=0)
    effective_play_events: int = Field(ge=0)


class ArchiveEncounterBin(StrictArchiveModel):
    key: Literal["same_day", "days_1_7", "days_8_30", "days_31_90", "days_90_plus"]
    entities: int = Field(ge=0)
    share_pct: float = Field(ge=0, le=100)


class ArchiveEncounterToSave(StrictArchiveModel):
    eligible_entities: int = Field(ge=0)
    no_observed_pre_save_play: int = Field(ge=0)
    bins: list[ArchiveEncounterBin] = Field(default_factory=list, min_length=5, max_length=5)
    examples: list[ArchiveEntityPreview] = Field(default_factory=list, max_length=5)


class ArchiveSymmetricWindow(StrictArchiveModel):
    window_days: Literal[30] = 30
    eligible_entities: int = Field(ge=0)
    before_events: int = Field(ge=0)
    after_events: int = Field(ge=0)
    more_before: int = Field(ge=0)
    equal: int = Field(ge=0)
    more_after: int = Field(ge=0)


class ArchiveReturnWindow(StrictArchiveModel):
    horizon_days: Literal[7, 30, 90, 365]
    eligible_entities: int = Field(ge=0)
    returned_entities: int = Field(ge=0)
    return_rate_pct: float | None = Field(default=None, ge=0, le=100)
    display_status: ArchiveMetricDisplayStatus


class ArchiveAlignedWeek(StrictArchiveModel):
    week_index: int = Field(ge=-4, le=12)
    eligible_entities: int = Field(ge=0)
    entities_with_play: int = Field(ge=0)
    effective_play_events: int = Field(ge=0)
    events_per_eligible: float = Field(ge=0)


class ArchiveRelationshipCounts(StrictArchiveModel):
    recent_active_saved: int = Field(ge=0)
    sleeping_saved: int = Field(ge=0)
    recently_saved_without_recent_play: int = Field(ge=0)
    saved_without_date: int = Field(ge=0)
    frequent_unsaved: int = Field(ge=0)
    unmatched_saved_tracks: int = Field(ge=0)


class ArchiveRelationshipMatrix(StrictArchiveModel):
    recent_window_days: Literal[90] = 90
    frequent_unsaved_min_plays: Literal[5] = 5
    counts: ArchiveRelationshipCounts
    recent_active_examples: list[ArchiveEntityPreview] = Field(default_factory=list, max_length=5)
    sleeping_examples: list[ArchiveEntityPreview] = Field(default_factory=list, max_length=5)
    frequent_unsaved_examples: list[ArchiveEntityPreview] = Field(
        default_factory=list, max_length=5
    )


class ArchiveCohortsResponse(StrictArchiveModel):
    schema_version: Literal["account_archive_cohorts_v1"] = "account_archive_cohorts_v1"
    content_version: Literal["account_archive_cohorts_v1_0"] = "account_archive_cohorts_v1_0"
    data_revision: str
    status: ArchiveChapterStatus
    filter_context: ArchiveFilterContext
    coverage: ArchiveCohortCoverage
    encounter_to_save: ArchiveEncounterToSave
    symmetric_30_day_window: ArchiveSymmetricWindow
    return_windows: list[ArchiveReturnWindow] = Field(
        default_factory=list, min_length=4, max_length=4
    )
    aligned_weeks: list[ArchiveAlignedWeek] = Field(default_factory=list)
    relationship_matrix: ArchiveRelationshipMatrix


class ArchiveReturnsCoverage(StrictArchiveModel):
    saved_tracks: int = Field(ge=0)
    matched_saved_tracks: int = Field(ge=0)
    unmatched_saved_tracks: int = Field(ge=0)
    canonical_saved_entities: int = Field(ge=0)
    dated_canonical_entities: int = Field(ge=0)
    invalid_added_dates: int = Field(ge=0)
    entities_with_effective_history: int = Field(ge=0)
    return_eligible_entities: int = Field(ge=0)
    effective_play_events: int = Field(ge=0)


class ArchiveReturnsSummary(StrictArchiveModel):
    gap_threshold_days: Literal[90] = 90
    return_episodes: int = Field(ge=0)
    returned_entities: int = Field(ge=0)
    multiple_return_entities: int = Field(ge=0)
    recent_30_day_return_entities: int = Field(ge=0)
    recent_90_day_return_entities: int = Field(ge=0)
    current_sleeping_entities: int = Field(ge=0)


class ArchiveReturnStory(StrictArchiveModel):
    track_name: str
    artist_name: str
    album_name: str | None = None
    cover_url: str | None = None
    deep_link: str | None = None
    added_date: str
    previous_play_at: str
    returned_at: str
    dormant_days: int = Field(ge=90)
    return_count: int = Field(ge=1)


class ArchiveSleepingStory(StrictArchiveModel):
    track_name: str
    artist_name: str
    album_name: str | None = None
    cover_url: str | None = None
    deep_link: str | None = None
    added_date: str
    last_play_at: str | None = None
    dormant_days: int = Field(ge=90)
    effective_plays: int = Field(ge=0)


class ArchiveReturnsResponse(StrictArchiveModel):
    schema_version: Literal["account_archive_returns_v1"] = "account_archive_returns_v1"
    content_version: Literal["account_archive_returns_v1_0"] = "account_archive_returns_v1_0"
    data_revision: str
    status: ArchiveChapterStatus
    filter_context: ArchiveFilterContext
    coverage: ArchiveReturnsCoverage
    summary: ArchiveReturnsSummary
    latest_returns: list[ArchiveReturnStory] = Field(default_factory=list, max_length=5)
    longest_returns: list[ArchiveReturnStory] = Field(default_factory=list, max_length=5)
    sleeping_recommendations: list[ArchiveSleepingStory] = Field(default_factory=list, max_length=5)


class ArchiveDiscoveryPeriod(StrictArchiveModel):
    first_search_at: str | None = None
    latest_search_at: str | None = None
    active_days: int = Field(ge=0)


class ArchiveDiscoveryCoverage(StrictArchiveModel):
    normalization_version: Literal["nfkc_casefold_ws_v1"] = "nfkc_casefold_ws_v1"
    burst_gap_minutes: Literal[5] = 5
    raw_search_rows: int = Field(ge=0)
    deduplicated_search_rows: int = Field(ge=0)
    invalid_timestamp_rows: int = Field(ge=0)
    unique_normalized_queries: int = Field(ge=0)
    search_bursts: int = Field(ge=0)
    interaction_records: int = Field(ge=0)
    interaction_bursts: int = Field(ge=0)


class ArchiveInteractionTypeCounts(StrictArchiveModel):
    track: int = Field(ge=0)
    artist: int = Field(ge=0)
    album: int = Field(ge=0)
    playlist: int = Field(ge=0)
    show: int = Field(ge=0)
    episode: int = Field(ge=0)
    other: int = Field(ge=0)


class ArchiveDiscoveryFunnel(StrictArchiveModel):
    display_status: Literal["count_only", "unavailable"]
    playback_window_minutes: Literal[60] = 60
    save_window_days: Literal[30] = 30
    track_interaction_bursts: int = Field(ge=0)
    mapped_track_interaction_bursts: int = Field(ge=0)
    played_within_1h_bursts: int = Field(ge=0)
    currently_saved_within_30d_bursts: int = Field(ge=0)


class ArchiveDiscoveryWeekdayPoint(StrictArchiveModel):
    weekday: int = Field(ge=0, le=6)
    bursts: int = Field(ge=0)


class ArchiveDiscoveryHourPoint(StrictArchiveModel):
    hour: int = Field(ge=0, le=23)
    bursts: int = Field(ge=0)


class ArchiveDiscoveryTrackPreview(StrictArchiveModel):
    track_name: str
    artist_name: str
    album_name: str | None = None
    cover_url: str | None = None
    deep_link: str | None = None
    interaction_at: str
    played_at: str
    added_date: str


class ArchiveDiscoveryResponse(StrictArchiveModel):
    schema_version: Literal["account_archive_discovery_v1"] = "account_archive_discovery_v1"
    content_version: Literal["account_archive_discovery_v1_0"] = "account_archive_discovery_v1_0"
    data_revision: str
    status: ArchiveChapterStatus
    filter_context: ArchiveFilterContext
    period: ArchiveDiscoveryPeriod
    coverage: ArchiveDiscoveryCoverage
    interaction_types: ArchiveInteractionTypeCounts
    funnel: ArchiveDiscoveryFunnel
    weekday_distribution: list[ArchiveDiscoveryWeekdayPoint] = Field(
        default_factory=list, min_length=7, max_length=7
    )
    hour_distribution: list[ArchiveDiscoveryHourPoint] = Field(
        default_factory=list, min_length=24, max_length=24
    )
    observed_saved_examples: list[ArchiveDiscoveryTrackPreview] = Field(
        default_factory=list, max_length=5
    )


ArchiveLibraryEntityType = Literal["tracks", "albums", "artists", "playlists"]
ArchiveLibrarySort = Literal["recent", "oldest", "name", "artist", "tracks"]


class ArchiveLibraryTrackItem(StrictArchiveModel):
    entity_type: Literal["track"] = "track"
    item_key: str
    track_name: str
    artist_name: str
    album_name: str | None = None
    added_date: str | None = None
    cover_url: str | None = None
    deep_link: str | None = None


class ArchiveLibraryAlbumItem(StrictArchiveModel):
    entity_type: Literal["album"] = "album"
    item_key: str
    album_name: str
    artist_name: str
    cover_url: str | None = None
    deep_link: str | None = None


class ArchiveLibraryArtistItem(StrictArchiveModel):
    entity_type: Literal["artist"] = "artist"
    item_key: str
    artist_name: str
    cover_url: str | None = None
    deep_link: str | None = None


class ArchiveLibraryPlaylistTrackPreview(StrictArchiveModel):
    track_name: str
    artist_name: str


class ArchiveLibraryPlaylistItem(StrictArchiveModel):
    entity_type: Literal["playlist"] = "playlist"
    item_key: str
    playlist_id: int = Field(ge=1)
    playlist_name: str
    last_modified_date: str | None = None
    track_count: int = Field(ge=0)
    follower_count: int = Field(ge=0)
    preview_tracks: list[ArchiveLibraryPlaylistTrackPreview] = Field(
        default_factory=list, max_length=3
    )


ArchiveLibraryItem = Union[
    ArchiveLibraryTrackItem,
    ArchiveLibraryAlbumItem,
    ArchiveLibraryArtistItem,
    ArchiveLibraryPlaylistItem,
]


class ArchiveLibraryPageResponse(StrictArchiveModel):
    schema_version: Literal["account_archive_library_v1"] = "account_archive_library_v1"
    content_version: Literal["account_archive_library_v1_0"] = "account_archive_library_v1_0"
    data_revision: str
    entity_type: ArchiveLibraryEntityType
    page: int = Field(ge=1)
    limit: int = Field(ge=1, le=50)
    total: int = Field(ge=0)
    total_pages: int = Field(ge=0)
    sort: ArchiveLibrarySort
    search_applied: bool
    items: list[ArchiveLibraryItem] = Field(default_factory=list)


class ArchiveMediaObservationWindow(StrictArchiveModel):
    first_play_at: str | None = None
    latest_play_at: str | None = None


class ArchivePodcastShowPreview(StrictArchiveModel):
    show_name: str
    effective_events: int = Field(ge=0)
    effective_ms: int = Field(ge=0)


class ArchivePodcastSummary(StrictArchiveModel):
    source_rows: int = Field(ge=0)
    effective_events: int = Field(ge=0)
    effective_ms: int = Field(ge=0)
    unique_shows: int = Field(ge=0)
    active_months: int = Field(ge=0)
    returning_shows: int = Field(ge=0)
    first_effective_at: str | None = None
    latest_effective_at: str | None = None
    top_shows: list[ArchivePodcastShowPreview] = Field(default_factory=list, max_length=3)


class ArchiveVideoSummary(StrictArchiveModel):
    source_rows: int = Field(ge=0)
    effective_events: int = Field(ge=0)
    effective_ms: int = Field(ge=0)
    unique_tracks: int = Field(ge=0)
    active_days: int = Field(ge=0)
    first_effective_at: str | None = None
    latest_effective_at: str | None = None


class ArchiveAudioVideoComparison(StrictArchiveModel):
    audio_effective_events: int = Field(ge=0)
    audio_effective_ms: int = Field(ge=0)
    video_effective_events: int = Field(ge=0)
    video_effective_ms: int = Field(ge=0)


class ArchiveOtherMediaResponse(StrictArchiveModel):
    schema_version: Literal["account_archive_other_media_v1"] = "account_archive_other_media_v1"
    content_version: Literal["account_archive_other_media_v1_0"] = (
        "account_archive_other_media_v1_0"
    )
    data_revision: str
    status: ArchiveChapterStatus
    filter_context: ArchiveFilterContext
    observation_window: ArchiveMediaObservationWindow
    podcast: ArchivePodcastSummary
    video: ArchiveVideoSummary
    audio_video_comparison: ArchiveAudioVideoComparison

File: frontend/src/types/accountArchive.ts
export type ArchiveStatus = 'ready' | 'partial' | 'empty'
export type ArchiveChapterStatus = 'available' | 'partial' | 'unavailable'
export type ArchiveCapabilityStatus = 'available' | 'partial' | 'unavailable'
export type ArchiveDisplayStatus = 'stable_rate' | 'count_only' | 'unavailable'
export type ArchiveSectionKey =
  | 'cover'
  | 'journey'
  | 'cohorts'
  | 'relationships'
  | 'returns'
  | 'discovery'
  | 'library'
  | 'other-media'

export interface ArchiveFeaturedItem {
  role: 'first_saved' | 'latest_saved' | 'oldest_release' | 'newest_release'
  track_name: string
  artist_name: string
  album_name: string | null
  added_date: string | null
  release_date: string | null
  cover_url: string | null
  deep_link: string | null
}

export interface ArchiveOverview {
  schema_version: 'account_archive_v1'
  content_version: 'account_archive_v1_0'
  data_revision: string
  status: ArchiveStatus
  counts: {
    saved_tracks: number
    saved_albums: number
    saved_artists: number
    saved_shows: number
    playlists: number
    playlist_items: number
  }
  coverage: {
    saved_tracks_with_date: number
    saved_tracks_with_date_pct: number
    saved_tracks_linked_to_history: number
    saved_tracks_linked_to_history_pct: number
    saved_tracks_with_known_duration: number
    saved_tracks_with_known_duration_pct: number
    known_duration_ms: number
  }
  period: {
    first_saved_at: string | null
    latest_saved_at: string | null
    first_play_date: string | null
    latest_play_date: string | null
  }
  date_provenance: { oauth: number; manual: number; legacy: number; missing: number }
  capabilities: {
    collection_browse: ArchiveCapabilityStatus
    collection_timeline: ArchiveCapabilityStatus
    playback_cross_analysis: ArchiveCapabilityStatus
  }
  featured_items: ArchiveFeaturedItem[]
}

export interface ArchiveFilterContext {
  context_version: 'account_archive_filter_v2'
  min_ms: number
  music_only: true
  merge_enabled: boolean
  dynamic_threshold: boolean
  max_merge_gap_minutes: number
  merge_level: number
  timezone: 'Asia/Shanghai'
  first_play_at: string | null
  latest_play_at: string | null
  latest_play_date: string | null
  source_revision: string
  track_group_revision: string
  filter_fingerprint: string
}

export interface ArchiveEntityPreview {
  track_name: string
  artist_name: string
  album_name: string | null
  cover_url: string | null
  deep_link: string | null
  added_date: string | null
  first_play_at: string | null
  last_play_at: string | null
  effective_plays: number
  days_to_save: number | null
}

export interface ArchiveCollectionMilestone {
  ordinal: number
  track_name: string
  artist_name: string
  album_name: string | null
  added_date: string
  cover_url: string | null
  deep_link: string | null
}

export interface ArchiveJourney {
  schema_version: 'account_archive_journey_v2'
  content_version: 'account_archive_journey_v2_0'
  data_revision: string
  status: ArchiveChapterStatus
  filter_context: ArchiveFilterContext
  coverage: {
    saved_tracks: number
    saved_tracks_with_date: number
    invalid_added_dates: number
    saved_tracks_with_known_duration: number
    duration_coverage_pct: number
  }
  duration: {
    known_duration_ms: number
    release_year_start: number | null
    release_year_end: number | null
  }
  annual_growth: ArchiveGrowthPoint[]
  quarterly_growth: ArchiveGrowthPoint[]
  milestones: ArchiveCollectionMilestone[]
}

export interface ArchiveGrowthPoint {
  period: string
  year: number
  quarter: number | null
  saved_tracks: number
  cumulative_saved_tracks: number
}

export interface ArchiveCohorts {
  schema_version: 'account_archive_cohorts_v2'
  content_version: 'account_archive_cohorts_v2_0'
  data_revision: string
  status: ArchiveChapterStatus
  filter_context: ArchiveFilterContext
  coverage: {
    saved_tracks: number
    matched_saved_tracks: number
    unmatched_saved_tracks: number
    canonical_saved_entities: number
    dated_canonical_entities: number
    invalid_added_dates: number
    effective_play_events: number
  }
  encounter_to_save: {
    eligible_entities: number
    no_observed_pre_save_play: number
    bins: Array<{
      key: 'same_day' | 'days_1_7' | 'days_8_30' | 'days_31_90' | 'days_90_plus'
      entities: number
      share_pct: number
    }>
    examples: ArchiveEntityPreview[]
  }
  symmetric_30_day_window: {
    window_days: 30
    eligible_entities: number
    before_events: number
    after_events: number
    more_before: number
    equal: number
    more_after: number
  }
  return_windows: Array<{
    horizon_days: 7 | 30 | 90 | 365
    eligible_entities: number
    returned_entities: number
    return_rate_pct: number | null
    display_status: ArchiveDisplayStatus
  }>
  vitality_metrics: Array<{
    key: 'within_7d' | 'days_8_30' | 'after_180d' | 'after_365d'
    start_day: number
    end_day: number | null
    eligible_entities: number
    returned_entities: number
    return_rate_pct: number | null
    display_status: ArchiveDisplayStatus
  }>
  aligned_weeks: Array<{
    week_index: number
    eligible_entities: number
    entities_with_play: number
    effective_play_events: number
    events_per_eligible: number
  }>
  relationship_matrix: {
    recent_window_days: 90
    frequent_unsaved_min_plays: 5
    counts: {
      recent_active_saved: number
      sleeping_saved: number
      recently_saved_without_recent_play: number
      saved_without_date: number
      frequent_unsaved: number
      unmatched_saved_tracks: number
    }
    recent_active_examples: ArchiveEntityPreview[]
    sleeping_examples: ArchiveEntityPreview[]
    frequent_unsaved_examples: ArchiveEntityPreview[]
  }
}

export interface ArchiveReturnStory {
  track_name: string
  artist_name: string
  album_name: string | null
  cover_url: string | null
  deep_link: string | null
  added_date: string
  previous_play_at: string
  returned_at: string
  dormant_days: number
  return_count: number
}

export interface ArchiveSleepingStory {
  track_name: string
  artist_name: string
  album_name: string | null
  cover_url: string | null
  deep_link: string | null
  added_date: string
  last_play_at: string | null
  dormant_days: number
  effective_plays: number
}

export interface ArchiveReturns {
  schema_version: 'account_archive_returns_v1'
  content_version: 'account_archive_returns_v1_0'
  data_revision: string
  status: ArchiveChapterStatus
  filter_context: ArchiveFilterContext
  coverage: {
    saved_tracks: number
    matched_saved_tracks: number
    unmatched_saved_tracks: number
    canonical_saved_entities: number
    dated_canonical_entities: number
    invalid_added_dates: number
    entities_with_effective_history: number
    return_eligible_entities: number
    effective_play_events: number
  }
  summary: {
    gap_threshold_days: 90
    return_episodes: number
    returned_entities: number
    multiple_return_entities: number
    recent_30_day_return_entities: number
    recent_90_day_return_entities: number
    current_sleeping_entities: number
  }
  latest_returns: ArchiveReturnStory[]
  longest_returns: ArchiveReturnStory[]
  sleeping_recommendations: ArchiveSleepingStory[]
}

export interface ArchiveDiscovery {
  schema_version: 'account_archive_discovery_v1'
  content_version: 'account_archive_discovery_v1_0'
  data_revision: string
  status: ArchiveChapterStatus
  filter_context: ArchiveFilterContext
  period: { first_search_at: string | null; latest_search_at: string | null; active_days: number }
  coverage: {
    normalization_version: 'nfkc_casefold_ws_v1'
    burst_gap_minutes: 5
    raw_search_rows: number
    deduplicated_search_rows: number
    invalid_timestamp_rows: number
    unique_normalized_queries: number
    search_bursts: number
    interaction_records: number
    interaction_bursts: number
  }
  interaction_types: {
    track: number
    artist: number
    album: number
    playlist: number
    show: number
    episode: number
    other: number
  }
  funnel: {
    display_status: 'count_only' | 'unavailable'
    playback_window_minutes: 60
    save_window_days: 30
    track_interaction_bursts: number
    mapped_track_interaction_bursts: number
    played_within_1h_bursts: number
    currently_saved_within_30d_bursts: number
  }
  weekday_distribution: Array<{ weekday: number; bursts: number }>
  hour_distribution: Array<{ hour: number; bursts: number }>
  observed_saved_examples: Array<
    ArchiveEntityPreview & { interaction_at: string; played_at: string; added_date: string }
  >
}

export type ArchiveLibraryEntityType = 'tracks' | 'albums' | 'artists' | 'playlists'
export type ArchiveLibrarySort = 'recent' | 'oldest' | 'name' | 'artist' | 'tracks'

interface ArchiveLibraryBaseItem {
  item_key: string
}

export interface ArchiveLibraryTrackItem extends ArchiveLibraryBaseItem {
  entity_type: 'track'
  track_name: string
  artist_name: string
  album_name: string | null
  added_date: string | null
  cover_url: string | null
  deep_link: string | null
}

export interface ArchiveLibraryAlbumItem extends ArchiveLibraryBaseItem {
  entity_type: 'album'
  album_name: string
  artist_name: string
  cover_url: string | null
  deep_link: string | null
}

export interface ArchiveLibraryArtistItem extends ArchiveLibraryBaseItem {
  entity_type: 'artist'
  artist_name: string
  cover_url: string | null
  deep_link: string | null
}

export interface ArchiveLibraryPlaylistItem extends ArchiveLibraryBaseItem {
  entity_type: 'playlist'
  playlist_id: number
  playlist_name: string
  last_modified_date: string | null
  track_count: number
  follower_count: number
  preview_tracks: Array<{ track_name: string; artist_name: string }>
}

export type ArchiveLibraryItem =
  | ArchiveLibraryTrackItem
  | ArchiveLibraryAlbumItem
  | ArchiveLibraryArtistItem
  | ArchiveLibraryPlaylistItem

export interface ArchiveLibraryPage {
  schema_version: 'account_archive_library_v1'
  content_version: 'account_archive_library_v1_0'
  data_revision: string
  entity_type: ArchiveLibraryEntityType
  page: number
  limit: number
  total: number
  total_pages: number
  sort: ArchiveLibrarySort
  search_applied: boolean
  items: ArchiveLibraryItem[]
}

export interface ArchiveOtherMedia {
  schema_version: 'account_archive_other_media_v2'
  content_version: 'account_archive_other_media_v2_0'
  data_revision: string
  status: ArchiveChapterStatus
  filter_context: ArchiveFilterContext
  observation_window: { first_play_at: string | null; latest_play_at: string | null }
  podcast: {
    source_rows: number
    effective_events: number
    effective_ms: number
    unique_shows: number
    active_months: number
    returning_shows: number
    first_effective_at: string | null
    latest_effective_at: string | null
    top_shows: Array<{
      show_name: string
      publisher: string | null
      cover_url: string | null
      effective_events: number
      effective_ms: number
    }>
  }
  video: {
    source_rows: number
    effective_events: number
    effective_ms: number
    unique_tracks: number
    active_days: number
    first_effective_at: string | null
    latest_effective_at: string | null
    top_tracks: Array<{
      track_name: string
      artist_name: string
      album_name: string | null
      cover_url: string | null
      deep_link: string | null
      effective_events: number
      effective_ms: number
    }>
  }
  audio_video_comparison: {
    audio_effective_events: number
    audio_effective_ms: number
    video_effective_events: number
    video_effective_ms: number
  }
}
